lib/dal/src/layer_db_types/content_types.rs

```diff
@@ -435,48 +451,66 @@ pub struct SchemaContentV1 {
-                    version: v1.name.to_owned(),
-                    display_name,
-                    category: v1.category,
-                    color: v1.color,
-                    component_type: v1.component_type,
-                    link: v1.link,
-                    description: v1.description,
-                    asset_func_id: v1.asset_func_id,
-                    finalized_once: v1.finalized_once,
-                    is_builtin: v1.is_builtin,
-                    is_locked: true,
-                })
-            }
-            later => later,
-        };
+        let mut working_content = self;
+        loop {
+            working_content = match working_content {
+                SchemaVariantContent::V1(v1) => {
+                    let display_name = if let Some(display_name) = v1.display_name {
+                        display_name
+                    } else {
+                        let schema = SchemaVariant::schema_for_schema_variant_id(ctx, id)
+                            .await
+                            .map_err(|e| ContentTypeError::SchemaVariantContent(e.to_string()))?;
+                        schema.name
+                    };
+
+                    SchemaVariantContent::V2(SchemaVariantContentV2 {
+                        timestamp: v1.timestamp,
+                        ui_hidden: v1.ui_hidden,
+                        version: v1.name.to_owned(),
+                        display_name,
+                        category: v1.category,
+                        color: v1.color,
+                        component_type: v1.component_type,
+                        link: v1.link,
+                        description: v1.description,
+                        asset_func_id: v1.asset_func_id,
+                        finalized_once: v1.finalized_once,
+                        is_builtin: v1.is_builtin,
+                        is_locked: true,
+                    })
+                }
+                SchemaVariantContent::V2(v2) => SchemaVariantContent::V3(SchemaVariantContentV3 {
+                    timestamp: v2.timestamp,
+                    ui_hidden: v2.ui_hidden,
+                    version: v2.version,
+                    display_name: v2.display_name,
+                    category: v2.category,
+                    color: v2.color,
+                    component_type: v2.component_type,
+                    link: v2.link,
+                    description: v2.description,
+                    asset_func_id: v2.asset_func_id,
+                    finalized_once: v2.finalized_once,
+                    is_builtin: v2.is_builtin,
+                }),
+                SchemaVariantContent::V3(_) => break,
+            };
+        }
 
         // extract latest
-        let latest = match at_least_v2 {
-            SchemaVariantContent::V2(v2) => v2,
+        let latest = match working_content {
+            SchemaVariantContent::V3(v3) => v3,
             _ => unreachable!(),
         };
```

Comment:
Feels like we could refactor to remove the blank (😱) unreachable. This is nit though.

Reply:
it may make sense to return an error here, although we know this is actually unreachable, so I'm ok with it.

Reply:
On a second thought: we should implement a proc macro at some point to handle this gradual upgrade process so that we can always be sure it's unreachable.